### ISSUE TITLE
error: None of the specified formats were available..."

### ISSUE THREAD

**Issue by jchhenderson** [Ascoware/get-iplayer-automator#95]
Hi,

OS: 		MacOS 10.13.2
iPlayer:	1.9.13

When downloading ITV "Midsomer Murders - fit for Murder", I get the following error:
___
	None of the specified formats were available.
	Please go to preferences -> Other Formats and add at least one download format to the list"
___
 
However, I had 3 formats selected (Very High, High, Standard) and previous episodes have downloaded just fine.  Just in case, I added all the other formats as well, but got the same error.

The log shows the following:
___
	ITV Download (ID=1/7317/0008#006): Downloading Midsomer Murders - Fit For Murder - 
	1/7317/0008#006
	INFO: Preparing Request for Auth Info
	INFO: Requesting Metadata.
	ITV Download (ID=1/7317/0008#006): ERROR: None of the modes in your download format list are 	available for this show. Try adding more modes if possible.
	AppController: Downloads Finished
___

Any ideas?
thanks,
James.

**Comment by GainfulShrimp** [comment]
+1.  Same issue with Through the Keyhole s05e01

**Comment by approtech** [comment]
This seems to be a growing issue with ITV; Issues 87 & 94 relate to the same thing.

I currently have the issue with three programmes:
The Protectors - Series 1 - Episode 10
The Protectors - Series 1 - Episode 13
The Widower - Series 1 - Episode 2

In each case, episodes of the same series before and after the problem episodes downloaded perfectly. Obviously something ITV are doing with certain episodes is breaking GiA.

The key part of the log is:

DEBUG: Metadata response: <s:Envelope xmlns:s="http://schemas.xmlsoap.org/soap/envelope/"><s:Body><GetPlaylistResponse xmlns="http://tempuri.org/"><GetPlaylistResult><Fault xmlns=""><faultcode>InvalidEntity</faultcode><faultstring>Video has no renditions for platform DotCom</faultstring></Fault></GetPlaylistResult></GetPlaylistResponse></s:Body></s:Envelope>

This appears for each episode.

**Comment by skovatch** [comment]
This is not a unique problem to GiA... youtube-dl is having similar issues. See https://github.com/rg3/youtube-dl/issues/14944

**Comment by mothyinuk** [comment]
Just found the same issue with Cilla episode 2 on ITV. Episode 1 & 3 downloaded OK.
Is there nothing that can be done but just check back with ITV to see if they have done something?

**Comment by ffooky** [comment]
I sometimes get this with episodes of Corontion Street. I usually find that those particular episodes won't play correctly on ITV Hub itself, often switching to another episode or the beginning of the first part of the same one. You can manually scrub to the correct part (as long as you've sat through the adverts) but I'm sure this glitch us what causes the failure in GiA.

Get Outlook for Android<https://aka.ms/ghei36>

________________________________

**Comment by rabrowne** [comment]
Same issue with Undeniable episode 1 on ITV.  Episode 2 downloaded ok. Episode 1 seems to play ok on ITV Hub though!

**Comment by jchhenderson** [comment]
@rabrowne yes, they play fine on ITV his for me too.  Only one day left for me to download my particular episode so pretty sure it's too late now.

**Comment by skovatch** [comment]
I can try to look at this over the holiday; I'll have some free time. Based on what I saw on youtube-dl, there's a possibility of downloading the non-commercial parts that show on the web site, and re-assembling them.

It's not like GiA is assembling a wrong URI -- when the URI is fetched, ITV reports back nothing is available.

**Comment by skovatch** [comment]
Here's what I think I know... ITV apparently makes multiple versions available. GiA attempts to fetch the MP4 directly via RTMPDump and ITV's media player. Sometimes, though, ITV doesn't make those files available. 

"But", you may say, "it's playable on the web page!" Well, there is another path to the video files that resembles what BBC is doing for streaming content. By following a different URI in the show page you can get a different JSON response that leads to a collection of MPEG pieces that can be reassembled.

This is how Downie is able to fetch a show when GiA cannot. I will investigate how I can go about replicating this in GiA. It may take a while but I think I see how it will work.

**Comment by murburz** [comment]
I was having the same problem and am comforted by your posts @skovatch. Thanks for all the work!

**Comment by approtech** [comment]
That would be a Christmas present! Many thanks for your continuing commitment to GiA, @skovatch

**Comment by hugofalkman** [comment]
@skovatch it seems the problem is hit in the metadata, could an mp4 be available but not the metadata? Verbose log for the new Maigret last night:

ITV Download (ID=2/4244/0004#002): Downloading Maigret - Maigret in Montmartre - 2/4244/0004#002
INFO: Preparing Request for Auth Info
DEBUG: Metadata URL: http://mercury.itv.com/PlaylistService.svc
INFO: Requesting Metadata.
DEBUG: Metadata response status code: 200
DEBUG: Metadata response: <s:Envelope xmlns:s="http://schemas.xmlsoap.org/soap/envelope/"><s:Body><GetPlaylistResponse xmlns="http://tempuri.org/"><GetPlaylistResult><Fault xmlns=""><faultcode>InvalidEntity</faultcode><faultstring>Video production has no renditions for DotCom platform.</faultstring></Fault></GetPlaylistResult></GetPlaylistResponse></s:Body></s:Envelope>
DEBUG: Metadata processed: seriesName=(null) dateString=(null) episodeName=(No Episode Name) episodeNumber=2 self.thumbnailURL=(null) subtitleURL=(null) authURL=(null)
DEBUG: Retrieving Playpath
DEBUG: Parsing MediaFile entries
DEBUG: Searching for itvRate match
DEBUG: Searching for bitrate match
ITV Download (ID=2/4244/0004#002): ERROR: None of the modes in your download format list are available for this show. Try adding more modes if possible.
AppController: Downloads Finished

**Comment by naq59** [comment]
Also noticed this on Maigret today. @skovatch - would be very warmly appreciated if could find a fix when you have the time. At least it seems to be possible to watch online.

**Comment by Lordsnooty05** [comment]
naq59 I agree, I'm also having terrible problems with the latest episode of Maigret wouldn't download at all. In fact, I'm having problems with most ITV programs. I'm sure Scott will sort it out.......Scott Happy New Year and have a restful time before you sort this problem out

**Comment by rabrowne** [comment]
Agree with the above, Christmas episode of Victoria on ITV also failing. :-(
Thank you Scott for your time and support of this great app. Wishing you a happy new year.

**Comment by MalArt09** [comment]
Yes... so great of you to take on the workload Scott (and others)... thanks for all you do.... but.... as Dale said in Flash Gordon... “ but [Scott] you only have 24 [days] to save the world!”.. as I would so like to save the Maigret episode before it expires! :)

**Comment by WilliamOfGaunt** [comment]
Possibly an extra data point: it may only be new programmes that are affected, because for some reason I succeeded in downloading today’s episode of Midsomer Murders, first broadcast 4 years ago...

Like the others, I’m extremely grateful to Scott for his continued maintenance of this invaluable app.
-- 
Bruce

**Comment by approtech** [comment]
I think it's more random - I've been downloading 'The Protectors' and had issues with Episodes 10, 13, 15 & 16 - but the ones in between (and Episode 17 shown today) all work fine.

**Comment by Themanwithoutaplan** [comment]
Until a fix becomes available you should be able to use [Downie](https://trial.charliemonroe.net/downie/download.php) to get stuff from ITV. Nice app also comes with browser plugins and a free trail period to see if it works for you.

**Comment by hugofalkman** [comment]
Downie doesn't seem to pick up subtitles from ITV. @skovatch would you be able to improve on the Downie technique to also download the subtitles?

**Comment by Themanwithoutaplan** [comment]
https://www.podnapisi.net/ is usually a good place to get subtitles is they aren't with the download but nothing yet for "Maigret in Montmartre".

**Comment by MalArt09** [comment]
Thanks **Themanwithoutaplan** for the heads up about Downie! I will always want GIA as my main tool but, alternatives are welcome in a crisis. I still hope GIA will have an update soon.
In the days of DVD recorders and TV's that record to a USB output, you would think that they would just be consistant and standardise what they are doing. In the end, perhaps it's about the possibility of further revenue from box sets that they are thinking of ?

**Comment by Lordsnooty05** [comment]
Themanwithoutaplan Thanks for heads up for Downie. It works after a fashion. Certainly not as good as GIA (in my personal opinion) but it is good as a back up when GIA isn't working 100% It does seem to download the shows at a very good quality....do you know if it is HD?
Anyway, I'm not here to talk about Downie but to say that I will wait until Scott sorts out the problem, which I'm sure he will. 
Happy New Year to Scott and all here who contribute to this fabulous Application

**Comment by amlmd181** [comment]
Thanks in advance to everyone working on this problem.  GiA is the best app for those of us outside the UK who are addicted to British TV.  May I recommend "The Miniaturist" on BBC, whose shows are still downloadable.

**Comment by Vangelis66** [comment]
Happy Holidays to one and all :santa: !

@skovatch 

Please take a look at my comment [here](https://github.com/monsieurvideo/get-flash-videos/issues/221#issuecomment-353695704); **get-flash-videos** (**GFV**), very much like **GiA**, relies upon the availability of unencrypted RTMPE (Flash) audio-video streams (downloadable by RTMPDump) from ITVHub; those streams represent an older infrastructure, probably still kept in the background for contingency reasons... 
ITVHub now serves desktop browsers encrypted MPEG-DASH streams; these are, of course, no good for GiA's purposes... 

It would appear that ITVHub are now removing, beit in a random fashion, RTMPE streams from their production line; those missing streams (for the shows affected) is the root cause of the issue reported in this thread... 

The only viable fallback is making use of the unencrypted AppleHLS streams that ITVHub are currently providing, targeting mobile devices (and it's a mystery for how long they'll remain unencrypted, i.e. downloadable/usable...). 

A (serious?) re-write of GiA would have to be undertaken to include an HLS downloader for ITV; the itv.pm plugin (originally part of get_iplayer) only provides ITVHub searching functions, it's of no use currently for integration with the get_iplayer perl script, _which does have a native, pure perl, HLS downloader_ but that's only reserved for BBC audio/video HLS streams... 

The yt-dl dev in https://github.com/rg3/youtube-dl/commit/3fae11ac00 has patched his code so that the ITV HLS streams are properly detected and then passed on to pre-existing HLS downloading python code...

**@allothers**: 

Until GiA is updated to accommodate HLS streams in ITVHub, one alternative is the CLI tool [youtube-dl](https://rg3.github.io/youtube-dl/); this is a Python script, so it does work in Mac OS X... 
https://github.com/rg3/youtube-dl/blob/master/README.md#installation
(Homebrew install: `brew install youtube-dl`)
If you're not familiar with the utility, `youtube-dl -h` or 
https://github.com/rg3/youtube-dl/blob/master/README.md

Normally, a downloaded HLS stream would be in the MPEG-TS container (.ts file extension); if you do want an .mp4 file, a rudimentary tag and .srt subtitles, you need a recent build of ffmpeg installed, too... 
(added bonus: youtube-dl has a built-in mechanism to circumvent ITVHub's geo-block (but not iPlayer's one); yes, it works even from outside the UK ; but please, don't flaunt this much...).
 
@approtech 
`The Protectors, Series 1 - Episode 16` =>
https://www.itv.com/hub/the-protectors/ENT1315a0024

`youtube-dl --console-title --hls-prefer-native -c --no-part "https://www.itv.com/hub/the-protectors/ENT1315a0024" --write-sub --convert-subs=srt --add-metadata`
```
[ITV] ENT1315a0024: Downloading webpage
[ITV] ENT1315a0024: Downloading XML
[ITV] ENT1315a0024: Downloading JSON metadata
[ITV] ENT1315a0024: Downloading m3u8 information
[hlsnative] Downloading m3u8 manifest
[hlsnative] Total fragments: 123
[download] Destination: The Protectors - Series 1 - Episode 16-ENT1315a0024.mp4
[download] 100% of 284.64MiB in 04:57
[ffmpeg] Fixing malformed AAC bitstream in "The Protectors - Series 1 - Episode
16-ENT1315a0024.mp4"
[ffmpeg] Adding metadata to 'The Protectors - Series 1 - Episode 16-ENT1315a0024
.mp4'
[ffmpeg] There aren't any subtitles to convert
```
Alas, I wasn't aware beforehand that the subs were missing for that specific episode, but I left the command as is regardless, as a template for those episodes with subs...

Happy New Year!

**Comment by skovatch** [comment]
@Vangelis66 That commit link is helpful, thanks! He's basically doing what I will end up doing natively -- fetch those segments, combine them all into one and use ffmpeg to convert from DASH to an MP4 that can be viewed in iTunes. I'm guessing you might be on Windows(?) as youtube-dl delegates the HLS download to ffmpeg on macOS. This is also what Downie is doing, I think. 

Also, I don't expect to change the subtitle-fetching code, as that doesn't need to change.

**Comment by Vangelis66** [comment]
> and use ffmpeg to convert from **DASH** to an MP4

...mmh, rather "from MPEG-TS (container) to MP4 (container/file)"; probably using something like the following incantation: 
`ffmpeg -v 8 -stats -i foo.ts -c copy -bsf:a aac_adtstoasc foo.mp4`

> as youtube-dl delegates the HLS download to ffmpeg on macOS.

... AFAICT the above behaviour is not OS specific, but indeed the default one; HLS stream dumping is by default delegated to FFmpeg (provided it is installed, i.e. the python script knows where to find it). 

youtube-dl _does come_ with a native python HLS downloader, but the yt-dl devs still consider it to be experimental/not fully matured; as such, it isn't enabled by default (and in the case of live HLS streams, still dysfunctional). 

For on-demand HLS streams (the case here with ITVhub catchup), you can explicitly enable the native python downloader by issuing the `--hls-prefer-native` switch (see my "template" command in previous comment); sadly, one still is in need of FFmpeg for the lossless remux of the dumped stream to iTunes palatable MP4 container...

(Best wishes for the New Year!)

**Comment by whitedb** [comment]
Thanks Vangelis66, your tip worked for me.

**Comment by GarthRanzz** [comment]
I keep getting the error "Unable to extract params" when trying youtube-dl and using the template above. Any ideas?

**Comment by nitramlatep** [comment]
Both Dancing on Ice and The Voice work first time over the weekend so it is clearly a hit and miss situation with itv and how they have posted the program on the hub as indicated above in the thread.

**Comment by WilliamOfGaunt** [comment]
Confirmed – Vera downloaded fine.

**Comment by rabrowne** [comment]
Agree with last two comments; Next of Kin failed but Girlfriends Episode 1 downloaded fine.

**Comment by nitramlatep** [comment]
@rabrowne can you try Next of Kin again.  I have just downloaded it without problem.  I suspect at the time you tried they had not posted a version that GiA could recognise.

**Comment by Lordsnooty05** [comment]
I also downloaded it without any problem.....it worked a treat. Are you using the most up to date version of GIA  Version 1.9.13 (20171126001)?

**Comment by rabrowne** [comment]
@nitramlatep , @Lordsnooty05 downloaded fine as you say. I am using the latest version of GIA. Maybe  just need to be patient if a download fails and wait a couple of days :-). Thanks.

**Comment by Lordsnooty05** [comment]
Glad it worked....looks like a good show. There are problems with ITV and some of their shows but Scott knows about it and as far as I know, he is working on a solution but you need to be patient and if a show doesn't  work the first time then you need to wait for a day or so and try again. If that doesn't work then you could try 'Downie'....it works after a fashion but not as good as GIA but it is an alternative until Scott works on the next version http://software.charliemonroe.net/downie.php

**Comment by jelockwood** [comment]
For what it's worth the Piers Morgan - Donald Trump interview as per https://www.itv.com/hub/president-trump-the-piers-morgan-interview/2a5582a0001 also fails.

**Comment by rabrowne** [comment]
"The Widower" and "Prey", both 3 part series on ITV each have one episode that fails with formats not found! Episode 2 for The Widower and Episode 1 for Prey.
Have you had a chance to look at any solutions yet @skovatch? Thanks again for all your support for this very useful app.

**Comment by hugofalkman** [comment]
I use Downie whenever GiA fails downloading ITV. Got the owner of Downie to add subtitle support for ITV. Downie settings to add to iTunes in TV program category. Also set download directory to the same as for GiA.

It then winds up in iTunes under the heading Unknown TV programs. Right click on the program, choose TV program info, and add program name, series, and episode.

**Comment by skovatch** [comment]
Now fixed in v1.11.0. GiA now uses the same path that Downie and Youtube-DL do for getting ITV shows, but still performs the same tagging, thumbnail fetching, subtitles, and adding to iTunes that GiA has always supported.

**Comment by Vangelis66** [comment]
> (added bonus: youtube-dl has a built-in mechanism to circumvent ITVHub's geo-block
> (snip)
> yes, it works even from outside the UK ; but please, don't flaunt this much...).

... :disappointed: After the recent (week of 19/03/2018), more srtict, geo-fencing scheme imposed by ITVHub, only the older RTMPE (flash) streams remain accessible to **youtube-dl** from overseas, the (newer) HLS streams are now blocked at CDN level... :-1: 

  Given this development, it might be desirable (for overseas GiA users) to still keep the RTMPE streams as a means of ITV media retrieval in GiA (for the shows they are available), rather than switch exclusively to the geo-fenced HLS streams (which, however, are the solution for this issue...).
